relay-server/src/endpoints/store.rs

```diff
@@ -19,6 +19,70 @@ use crate::service::{ServiceApp, ServiceState};
 static PIXEL: &[u8] =
     b"GIF89a\x01\x00\x01\x00\x00\xff\x00,\x00\x00\x00\x00\x01\x00\x01\x00\x00\x02\x00;";
 
+/// Parses a full `Envelope` request body.
+fn parse_envelope(meta: RequestMeta, data: Bytes) -> Result<Envelope, BadStoreRequest> {
+    // Use `parse_request` here to ensure that we're merging available request headers into the
+    // envelope's headers.
+    let mut envelope =
+        Envelope::parse_request(data, meta).map_err(BadStoreRequest::InvalidEnvelope)?;
+
+    // Compared to raw event JSON, we assume that clients send more sanitized event payloads.
+    // However, clients do not have to send an event_type, which is required by Relay to
+    // emit proper outcomes and enforce rate limits.
+    if let Some(event) = envelope.items_mut().find(|i| i.ty() == ItemType::Event) {
+        if event.event_type().is_none() {
+            let minimal = common::minimal_event_from_json(&event.payload())?;
+            event.set_event_type(minimal.ty);
+        }
+    }
+
+    Ok(envelope)
+}
+
+/// Parses a JSON event body into an `Envelope`.
+fn parse_event(
+    meta: RequestMeta,
+    content_type: String,
+    mut data: Bytes,
+) -> Result<Envelope, BadStoreRequest> {
+    // Python clients are well known to send crappy JSON in the Sentry world.  The reason
+    // for this is that they send NaN and Infinity as invalid JSON tokens.  The code sentry
+    // server could deal with this but we cannot.  To work around this issue, we run a basic
+    // character substitution on the input stream but only if we detect a Python agent.
+    //
+    // This is done here so that the rest of the code can assume valid JSON.
+    let is_legacy_python_json = meta.client().map_or(false, |agent| {
+        agent.starts_with("raven-python/") || agent.starts_with("sentry-python/")
```

Comment:
The current Python SDK sends only valid JSON, so we don't have to apply this fix when envelopes are sent.